feat(cli): add job status information to queue-status command

gajeshbhat opened the pull request:
## Description

This PR enhances the existing `queue-status` command to display job status information alongside the current agent status, addressing the feature request in issue #116.

Users need a way to get a summary of job statuses in a queue without polling individual job IDs. When running many jobs in sequence, it's difficult to get a quick overview of job progress and identify active work.

The solution extends the existing `queue-status` command to show both agent and job status in unified output, using a simple `--verbose` flag for detailed job information.

## Resolved issues

Fixes #116

## Documentation

- Tutorial updated to include `queue-status` command in CLI help example
- Command help text clearly describes all options and functionality
- Existing documentation patterns and usage remain valid

## Web service API changes

No API changes required. Uses existing endpoints:
- `/v1/queues/{queue}/agents` for agent status information  
- `/v1/queues/{queue}/jobs` for job status information

## Tests

**How this PR was tested:**

1. **Unit Tests (50/50 passing):**
   ```bash
   uv run python -m pytest testflinger_cli/tests/test_cli.py -v
   ```

2. **Linting & Formatting:**
   ```bash
   uvx --with tox-uv tox -e lint
   ```

3. **Manual Testing:**
   ```bash
   # Test basic functionality
   testflinger-cli queue-status my-queue
   
   # Test verbose mode  
   testflinger-cli queue-status --verbose my-queue
   
   # Test JSON output
   testflinger-cli queue-status --json my-queue
   ```

**Test coverage includes:**
- Default output with enhanced job counts
- Verbose mode showing individual job details with timestamps
- JSON output with new job status keys
- Error handling for empty/nonexistent queues
- Backward compatibility validation

**Key features implemented:**
- **Non-breaking**: Existing behavior completely preserved
- **Unified output**: Always shows both agent and job status together  
- **Simple interface**: Single `--verbose` flag for detailed information
- **Smart categorization**: Jobs grouped as waiting/running/completed (cancelled ignored)
- **JSON compatibility**: New keys added while maintaining backward compatibility

**Example outputs:**

Default (enhanced):
```
Agents in queue: 1
Available:       1
Busy:            0
Offline:         0
Jobs waiting:    0
Jobs running:    0    # new
Jobs completed:  2    # new
```

Verbose (new):
```
Agents in queue: 1
Available:       1
Busy:            0
Offline:         0
Jobs waiting:    1
Jobs running:    0
Jobs completed:  2

Waiting:
  de153d8f-7d32-47d7-9a05-a20f2ef6bb35 - 2023-10-13 15:22:46

Completed:
  8b0bb52f-08d8-4671-b275-55d84a965f7c - 2023-10-13 15:22:30
```

This implementation addresses the core need from issue #116 while maintaining full compatibility with existing usage and following established project patterns, and also helps the project close #CERTTF-248 internally.

rene-oromtz commented on cli/testflinger_cli/__init__.py:
```diff
@@ -425,6 +441,18 @@ def agent_status(self):
 
     def queue_status(self):
         """Show the status of the agents in a specified queue."""
+        if self.args.jobs:
+            # New job-focused functionality
+            self._queue_status_jobs()
+        else:
+            # Original agent-focused functionality
+            self._queue_status_agents()
```
I like the idea you are proposing for separating the logic for getting status for agents/jobs! As a suggestion, I think we can make this `queue_status()` to do the output formatting and just use the helper functions to retrieve the agent/jobs status. 

e.g
```suggestion
    def queue_status(self):
        """Show agent and job status in a specified queue."""
            agents_status =  self._queue_status_agents() # This always returns all agents in queue and its status
            jobs_status = self._queue_status_jobs() # This returns "waiting" jobs by default or all jobs if verbose
            
            if self.args.json:
                output = json.dumps() # JSON parsing logic
            else:
                output = ("") # Human Readable Logic
             print(output) 
           
```

rene-oromtz commented on cli/testflinger_cli/__init__.py:
```diff
@@ -425,6 +441,18 @@ def agent_status(self):
+            # Original agent-focused functionality
+            self._queue_status_agents()
+
+    def _queue_status_agents(self):
+        """Show the status of the agents in a specified queue.
+
+        This is the original functionality.
```
This can be reformatted as follow based on current suggestions:
```suggestion
    def _queue_status_agents(self):
        """Retrieve the status of the agents in a specified queue.
```

rene-oromtz commented on cli/testflinger_cli/__init__.py:
```diff
@@ -496,6 +524,143 @@ def queue_status(self):
             )
         print(output)
 
+    def _queue_status_jobs(self):
+        """Show the status of jobs in a specified queue (new functionality)."""
```
This can be reformatted as follow based on current suggestions:
```suggestion
    def _queue_status_jobs(self):
        """Retrieve the status of jobs in a specified queue"""
```

rene-oromtz commented on cli/testflinger_cli/__init__.py:
```diff
@@ -496,6 +524,143 @@ def queue_status(self):
+        except UnknownStatusError as exc:
+            sys.exit(exc)
+
+        # Categorize jobs by status
+        job_counts = Counter()
+        incomplete_jobs = []
+        completed_jobs = []
```
I think the Counter can be simplified significantly (maybe without the summary) by just taking the following assumptions:
1. I believe the job status is just "complete" now
2. Assume all non "waiting", "complete", "cancelled" are jobs that are "running", this way you don't need to check all statuses individually.
3. I think we can ignore the cancelled jobs, this were made by the user request 
4. Add here the `--verbose` logic, if not verbose should only return the jobs that are waiting (default), also, I think we can add the timestamp for both default/verbose, this may can be of use and is already retrieved anyway. 

Example
Expected default return:
```
jobs_status = {
  "jobs_waiting": [
    {
      "job_id": "2394a56f-52d8-41a4-bf84-6e39ce4ad2de",
      "created_at": "2025-08-22T15:49:18.849Z"
    }
  ],
}
```

Verbose return:
```
jobs_status = {
  "jobs_waiting": [
    {
      "job_id": "2394a56f-52d8-41a4-bf84-6e39ce4ad2de",
      "created_at": "2025-08-22T15:49:18.849Z"
    }
  ],
  "jobs_completed": [
  ],
  "jobs_running": [
  ],
}
```

rene-oromtz commented on cli/testflinger_cli/__init__.py:
```diff
@@ -446,55 +466,120 @@ def queue_status(self):
+        output_data["jobs waiting"] = [
+            job["job_id"] for job in jobs_status["jobs_waiting"]
+        ]
+
+        if self.args.verbose:
+            # In verbose mode, include all job details
+            return json.dumps(output_data, indent=2)
         else:
-            agents = Counter(
-                (
-                    agent["status"]
-                    if agent["status"] in ("waiting", "offline")
-                    else "busy"
-                )
-                for agent in queue_status
+            # In non-verbose mode, include basic job info
+            return json.dumps(output_data, indent=2)
```
Glad you notice the "jobs waiting", however, we can drop this to follow the proper snake_case, no need to keep the backward compatibility as this was only used by this `queue_status`

Then it can be simplified as:
 ```
       if self.args.verbose:
            # In verbose mode, include all job details
            output_data.update(jobs_status)
        else:
            # In non-verbose mode, only include waiting jobs
            output_data["jobs_waiting"] = jobs_status["jobs_waiting"]
        return json.dumps(output_data, indent=2)
```

rene-oromtz commented on cli/testflinger_cli/__init__.py:
```diff
@@ -446,55 +466,120 @@ def queue_status(self):
-                f"Jobs waiting:    {len(jobs_queued)}"
-            )
-        print(output)
+        output_lines = [
+            f"Agents in queue: {agents.total()}",
+            f"Available:       {agents['waiting']}",
+            f"Busy:            {agents['busy']}",
+            f"Offline:         {agents['offline']}",
+            f"Jobs waiting:    {len(jobs_status['jobs_waiting'])}",
+            f"Jobs running:    {len(jobs_status['jobs_running'])}",
+            f"Jobs completed:  {len(jobs_status['jobs_completed'])}",
+        ]
+
+        if self.args.verbose:
+            # Add individual job details
+            for job_type, jobs in jobs_status.items():
+                if jobs:  # Only show if there are jobs
+                    job_type_display = (
+                        job_type.replace("jobs_", "").replace("_", " ").title()
+                    )
+                    output_lines.append(f"\n{job_type_display}:")
+                    for job in jobs:
+                        timestamp = self._format_timestamp(
+                            job.get("created_at", "")
+                        )
+                        output_lines.append(f"  {job['job_id']} - {timestamp}")
+
+        return "\n".join(output_lines)
```
Would you consider this instead? This to avoid the "double" job output ("Jobs waiting: count" + "Waiting: list")

```suggestion
        output_lines = [
            f"Agents in queue: {agents.total()}",
            f"Available:       {agents['waiting']}",
            f"Busy:            {agents['busy']}",
            f"Offline:         {agents['offline']}",
        ]

        if self.args.verbose:
            # Add individual job details
            for job_type, jobs in jobs_status.items():
                if jobs:  # Only show if there are jobs
                    job_type_display = job_type.replace("_", " ").title()
                    output_lines.append(f"\n{job_type_display}:")
                    for job in jobs:
                        timestamp = self._format_timestamp(
                            job.get("created_at", "")
                        )
                        output_lines.append(f"  {job['job_id']} - {timestamp}")
        else:
            output_lines.extend(
                [
                    f"Jobs waiting:    {len(jobs_status['jobs_waiting'])}",
                    f"Jobs running:    {len(jobs_status['jobs_running'])}",
                    f"Jobs completed:  {len(jobs_status['jobs_completed'])}",
                ]
            )

        return "\n".join(output_lines)
```

gajeshbhat replied:
Thank you for the suggestion. I updated it to use the extend keyword.

rene-oromtz reviewed (changes requested):
Thanks for adding the suggestions! Overall I'm liking the current approach, I just leave couple more suggestions and one important consideration regarding on the current MongoDB structure

gajeshbhat commented on cli/testflinger_cli/__init__.py:
```diff
@@ -446,55 +466,120 @@ def queue_status(self):
+            logger.debug("Unable to retrieve job data: %s", exc)
+            jobs_data = []
+
+        # Categorize jobs based on maintainer's suggestions
```
Updated! Thank you for pointing this out.

rene-oromtz commented on cli/testflinger_cli/__init__.py:
```diff
@@ -446,55 +466,121 @@ def queue_status(self):
+
+    def _format_human_output(self, agents_status, jobs_status):
+        """Format output for human reading."""
+        # Agent status (existing format)
```
Nitpick
```suggestion
        # Get agent status count
```

gajeshbhat replied:
Updated.

rene-oromtz commented on cli/testflinger_cli/__init__.py:
```diff
@@ -446,55 +466,121 @@ def queue_status(self):
+
+        return "\n".join(output_lines)
+
+    def _format_timestamp(self, timestamp_str):
+        """Format timestamp for human reading."""
+        if not timestamp_str:
+            return "Unknown"
+
+        try:
+            # Parse ISO format timestamp
+            dt = datetime.fromisoformat(timestamp_str.replace("Z", "+00:00"))
+            return dt.strftime("%Y-%m-%d %H:%M:%S")
+        except (ValueError, AttributeError):
+            return timestamp_str
```
This is just a suggestion, but since this can potentially be reused, it could be worth moving this logic to `helpers.py`. This module its already imported so it should be as easy as modify your previous call to `helpers.format_timestamp()` what do you think?

gajeshbhat replied:
Will move this to `helpers.py`.

rene-oromtz commented on cli/testflinger_cli/tests/test_cli.py:
```diff
@@ -745,6 +760,134 @@ def test_queue_status(capsys, requests_mock):
+
+    requests_mock.get(
+        URL + "/v1/queues/" + fake_queue + "/agents",
+        status_code=204,  # No content
```
Instead of comment, can you please modify to use HTTPStatus for readability ?
```suggestion
        status_code=HTTPStatus.NO_CONTENT
```

rene-oromtz commented on cli/testflinger_cli/tests/test_cli.py:
```diff
@@ -745,6 +760,134 @@ def test_queue_status(capsys, requests_mock):
+
+    requests_mock.get(
+        URL + "/v1/queues/" + fake_queue + "/agents",
+        status_code=404,  # Not found
```
Same comment as above
```suggestion
        status_code=HTTPStatus.NOT_FOUND
```

rene-oromtz reviewed (changes requested):
Thanks again for the changes! 
Just added final nitpicks/suggestions that I didn't catch before (sorry for that, this are the final comments I have). Functionally, the code looks good to me! 
Tested locally as well and its working as expected

gajeshbhat commented on cli/testflinger_cli/__init__.py:
```diff
@@ -446,55 +466,121 @@ def queue_status(self):
+        return json.dumps(output_data, indent=2)
+
+    def _format_human_output(self, agents_status, jobs_status):
+        """Format output for human reading."""
```
Same as above. Updated.

gajeshbhat commented:
> Thanks again for the changes! Just added final nitpicks/suggestions that I didn't catch before (sorry for that, these are the final comments I have). Functionally, the code looks good to me! Tested locally as well, and it's working as expected

Thanks again for taking a look at this @rene-oromtz .I have updated the changes you asked for. I just wanted to let you know that there's no need to be sorry. I am happy to always work with the maintainers to make sure my code submission meets the repository standards and does not introduce more technical debt. Please let me know if you have any further questions. Cheers.  :)

On a side note, I have also put up a [PR](https://github.com/canonical/testflinger/pull/787) for a [bug](https://github.com/canonical/testflinger/issues/438) that troubled me a lot in the past while running testflinger in production when agents crash due to intermittent network issues. I believe this bugfix would help many teams running TestFlinger. No rush/urgency on this. Take a look when you have free time. Cheers.
